couchbase_test.go

```diff
@@ -0,0 +1,694 @@
+package couchbase
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"os"
+	"testing"
+	"time"
+
+	docker "github.com/hashicorp/vault/helper/testhelpers/docker"
+	"github.com/hashicorp/vault/sdk/database/dbplugin"
+	"github.com/ory/dockertest"
+	dc "github.com/ory/dockertest/docker"
+)
+
+var containerInitialized bool = false
+var cleanup func() = func() {}
+var pre_6dot5 = false // check for Pre 6.5.0 Couchbase
+
+func prepareCouchbaseTestContainer(t *testing.T) (func(), string, int) {
+	if os.Getenv("COUCHBASE_HOST") != "" {
+		return func() {}, os.Getenv("COUCHBASE_HOST"), 0
+	}
+
+	if containerInitialized == true {
+		return cleanup, "localhost", 0
+	}
+
+	pool, err := dockertest.NewPool("")
+	if err != nil {
+		t.Fatalf("Failed to connect to docker: %s", err)
+	}
+
+	ro := &dockertest.RunOptions{
+		Repository:   "docker.io/fhitchen/vault-couchbase",
+		Tag:          "latest",
+		ExposedPorts: []string{"8091", "8092", "8093", "8094", "11207", "11210", "18091", "18092", "18093", "18094"},
```

Comment:
Ok, so here is the deal. I left it this way because I knew that you can remap the ports at docker run time, but inside the container the couchbase server does not know about this. After you connect to the remapped 8091 port it will send back the cluster port configuration, but it will be wrong and will not have the docker remapped port information. If you want to remap the ports because your network does PATing (Port Address Translation) you also need to edit the `/opt/couchbase/etc/couchbase/static_config` file and restart the server, something that is not possible in a stock docker image, or a custom one for that. That was in the 6.0 couchbase version, details [here](https://docs.couchbase.com/server/6.0/install/install-ports.html#map-custom-ports).

Now with 6.5.0 version they still say you have to [edit the static config file](https://docs.couchbase.com/server/6.5/install/install-ports.html#map-custom-ports) and restart the server and they warn that this purges all data. But the server nodes api provides a restful API to [Manage Alternative Addresses](https://docs.couchbase.com/server/current/rest-api/rest-set-up-alternate-address.html) which also seems to support port remapping but after I read it carefully, still no luck as it is only for 6 of the services, kv (Data Service), index (Index Service), n1ql (Query Service), fts (Search Service), cbas (Analytics Service), and eventing (Eventing Service). Not the HTTP management and memcached ports we need.

This command will bring back the cluster configuration.

`$ curl -v -u Administrator:Admin123 localhost:8091/pools/nodes `

So if I run a couchbase image with non-standard port mappings...

`docker run -d --name db -p 19091-19096:18091-18096 -p 20207:11208 -p 9091-9094:8091-8094 -p 20210:11210 fhitchen/vault-couchbase`

and run the command `$ curl -v -u Administrator:Admin123 localhost:9091/pools/nodes` in the response JSON, none of the ports have changed.
```JSON
         "configuredHostname" : "127.0.0.1:8091",
         "couchApiBase" : "http://172.17.0.2:8092/",
         "couchApiBaseHTTPS" : "https://172.17.0.2:18092/",


         "ports" : {
            "direct" : 11210,
            "distTCP" : 21100,
            "distTLS" : 21150,
            "httpsCAPI" : 18092,
            "httpsMgmt" : 18091
         },
```
This is what the gocb client will query and try and connect to the cluster and I think it will hang as the couchbase cluster will just send back the standard ports to the client.

What do you suggest?

Reply:
Hmm that's tricky. Is it possible to use `couchbase-cli setting-alternate-address`? [Docs](https://docs.couchbase.com/server/6.5/cli/cbcli/couchbase-cli-setting-alternate-address.html) and [forum post](https://forums.couchbase.com/t/make-the-ports-8092-8093-8094-and-8095-configurable-to-be-able-to-use-dockers-random-ports/22966) discussing it.

It would only be available for 6.5, and would be a little ugly to set up, but seems like it might do the trick? It's frustrating that the Go SDK client doesn't let you specify all the ports it should use upfront.

Reply:
I think that having to launch the couchbase container, get the external port mapping and then exec into the running container and use a cli tool to change the needed ports is going to be very brittle and as you pointed out, it will only work for the very latest version.

I think a simpler approach would be to use a docker-compose technique where you don't expose any ports outside of docker, just create a docker network, and hook your docker container with the tests in it into this software defined network would be the simplest and most robust approach. 

It looks like you can give the container a name in the [docker executor](https://circleci.com/docs/2.0/configuration-reference/#docker) which becomes it's node name. I am assuming that then from the go test container I can connect to the default ports just using http://couchbase etc. Have you any experience with this? This looks like a good [example](https://github.com/CircleCI-Public/circleci-demo-go/blob/master/.circleci/config.yml).

Reply:
Yeah I think docker networking is a better idea, although stepping back, I wonder if maybe it's best left as-is after all.

As long as this is the only test suite in the repo that runs the couchbase container (a reasonable constraint), the fact that the tests are running in a dedicated VM means this actually shouldn't be problematic with hardcoded ports. Given the difficulty and complexity of engineering around those fixed ports, it seems pragmatic to live with the constraint for now, and then reassess if it does cause issues down the line.

WDYT? Apologies for the wild goose chase on this one - I've been learning on the go about CircleCI and couchbase, so didn't have the full picture before.

Reply:
I would love the challenge of figuring out how to do this in Circleci, but it could take quite a while and I need to focus on the spring-cloud-vault integration. So lets go with this simple approach for now.